crates/pathfinder/src/sequencer.rs

```diff
@@ -29,6 +29,16 @@ fn block_hash_str(hash: BlockHashOrTag) -> (&'static str, Cow<'static, str>) {
     }
 }
 
+/// Starknet specific errors come with an internal server error HTTP staus code (500), so
+/// we have to treat them in a special way. Other HTTP errors are let through.
+fn check_for_error(response: &reqwest::Response) -> reqwest::Result<()> {
+    if response.status() == reqwest::StatusCode::INTERNAL_SERVER_ERROR {
+        return Ok(());
+    }
+
+    response.error_for_status_ref().map(|_| ())
+}
+
```

Comment:
Not a big fan of this function. It seems a bit off to pass a subset of "errors" through and then parse them separately.

Have you considered a function to parse the response, errors and all? I'll admit I haven't thought hard about this, so maybe I'm barking up the wrong tree.

Something like:
```Rust
fn parse_response<T>(response: &reqwest::Response) -> Result<T, SequencerError> {
    match response.status() {
        Status::OK => T::parse(response),
        Status::INTERNAL_SERVER_ERROR => serde_json::from_str::<StarknetError>(resp.as_str()),
        other => todo!("Map transport error? Or whatever goes here"),
    }
}
```

Reply:
The need here was that http client (400-499) or http server errors (501-599) should be propagated as `Err(SequencerError::TransportError)` while 500 should result in deserializing the body and only then being wrapped in `Err(SequencerError::StarknetError)`. As for other status codes - I assume that the logic of non-error statuses is correctly implemented by `reqwest` (and `hyper`).

So what this function does is:
1. effectively propagate `Err(SequencerError::TransportError)` for all except 500 error status codes with the error variant data being created by [`error_for_status_ref`](https://docs.rs/reqwest/latest/reqwest/struct.Response.html#method.error_for_status_ref)
2. propagate `Err(SequencerError::StarknetError)` if the body contains a valid Starknet error JSON object - this happens when a `sequencer::reply::[Block|Call|Code|...]Reply` enum is deserialized to its `Error` variant and then immediately `try_into()` is applied to extract either the valid data as `Ok(Block|Call|Code...)` (without the "Reply" suffix) or propagate as `Err(SequencerError::StarknetError)` if `try_into()` fails, for example:
```rust=
    /// Get block by hash.
    pub async fn block_by_hash(
        &self,
        block_hash: BlockHashOrTag,
    ) -> Result<reply::Block, SequencerError> {
        let (tag, hash) = block_hash_str(block_hash);
        // Attempt an HTTP GET
        let resp = reqwest::get(self.build_query("get_block", &[(tag, &hash)])).await?;
        // If status code is <400;499> or <501;599> return a `SequencerError::TransportError`
        check_for_error(&resp)?;
        // Get the text of the body
        let resp = resp.text().await?;
        // Deserialize either the proper reply variant of `reply::BlockReply` or its error variant and then
        // transform into either an Ok(`reply::Block`) or an `Err(SequencerError::StarknetError)`
        // respectively 
        serde_json::from_str::<reply::BlockReply>(resp.as_str())?.try_into()
    }

Ultimately I wanted to avoid parsing all HTTP status codes manually, which I think shouldn't be done by us, especially that there are special cases that are handled `reqwest` (or even `hyper`).
```

Reply:
We definitely want to avoid re-parsing the normal HTTP errors. What bothers me is that one can easily forget, overlook or delete the `check_for_error` call in-between it all. That's why I'm suggesting we bundle it together with the actual parsing code.

Basically instead of
```Rust
    /// Get block by hash.
    pub async fn block_by_hash(
        &self,
        block_hash: BlockHashOrTag,
    ) -> Result<reply::Block, SequencerError> {
        let (tag, hash) = block_hash_str(block_hash);
        // Attempt an HTTP GET
        let resp = reqwest::get(self.build_query("get_block", &[(tag, &hash)])).await?;
        // If status code is <400;499> or <501;599> return a `SequencerError::TransportError`
        check_for_error(&resp)?;
        // Get the text of the body
        let resp = resp.text().await?;
        // Deserialize either the proper reply variant of `reply::BlockReply` or its error variant and then
        // transform into either an Ok(`reply::Block`) or an `Err(SequencerError::StarknetError)`
        // respectively 
        serde_json::from_str::<reply::BlockReply>(resp.as_str())?.try_into()
}
```
you would have
```Rust
    /// Get block by hash.
    pub async fn block_by_hash(
        &self,
        block_hash: BlockHashOrTag,
    ) -> Result<reply::Block, SequencerError> {
        let (tag, hash) = block_hash_str(block_hash);
        // Attempt an HTTP GET
        let resp = reqwest::get(self.build_query("get_block", &[(tag, &hash)])).await?;
        parse::<reply::BlockReply>(resp)
}

// T will need to implement some Trait parse or something.
fn parse<T>(response: Response) -> Result<T, SequencerError> {
    match response.status() {
        Status::OK => T::parse(response),
        // We will likely need to amend this to return `StarknetError` or just the inner `SERVER_ERROR` if it doesn't map
        // (to handle cases where it is actually an unknown server code error -- can't recall if you handle those currently)
        Status::INTERNAL_SERVER_ERROR => serde_json::from_str::<StarknetError>(resp.as_str()),
        other => todo!("Map transport error? Or whatever goes here"),
    }
}
```

Reply:
Yes, I get it now, having to remember to stick this `check_for_error` fn each time is a big problem and is prone to error.

Reply:
I've got a single point of parsing fn ready, will push soon. Thanks to its introduction a lot of surplus code was removed :tada: